### PR TITLE
Functional handling for Fortran to Python

### PR DESCRIPTION
**Pull request by pratikbhd**
Additional test scripts added for function handling use cases. All scripts having "test" in the file name contain the additional function handling code. Once all enhancements and issues have been fixed, will merge this with the original scripts.

### CODE REVIEW

**Comment by pratikbhd** [comment]
That might be because I have not done any changes in the original files i.e. translate.py and pyTranslate.py. Instead, I made a copy of those files and renamed them test_translate.py and pytestTranslate.py. These are the files I worked on. 
Should I make those changes on the original files themselves and then commit them separately to see the diffs between them?

**Comment by pauldhein** [comment]
@pratikbhd I am not sure what @skdebray would prefer, but I would certainly prefer for you to make your changes to the original files so we can see diffs in these files overtime, and so that we do not clutter the repo with additional files. The system of using branches (perhaps even a branch from the program_analysis branch if necessary) coupled with pull requests and reviews will ensure that no unwanted changes to our main code occur. This should also give you agency to make whatever changes you wish to the files directly without worrying about their affects to other parts of delphi/program_analysis.

**Comment by skdebray** [comment]
Thanks.  I'll take a look as soon as I can (hopefully this afternoon).

At our meeting this afternoon, if we have a few mins it would be good to discuss the refactoring of the f2py translation process that was proposed over email last week.

File: delphi/program_analysis/autoTranslate/scripts/file_translate.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+            self.subroutine if subroutine == None else subroutine
+        )
+
+
+def loadFunction(root):
+    for element in root.iter():
+        if element.tag == "function":
+            functionList.append(element.attrib["name"])
+
+
```
Comment (skdebray) [inline delphi/program_analysis/autoTranslate/scripts/file_translate.py]: Pratik, is this code you wrote?  If so, can you please add a comment at the top stating what the function does?  Please do that for all of the functions you've written.  Thanks.

**Comment by adarshp** [comment]
Hi @pratikbhd,

I took a look at the code and made a couple of small cosmetic changes, so please be sure to pull them before working on it more.

Before I merge this branch into master, could you please do the following?

1. Remove automatically generated files, including but not limited to the following:
    - `custom_python_files/crop_yield.py`
    - `scripts/pgm_test/crop_yield.py`
    - `scripts/pgm_test/lambdas.py`
    - `scripts/pgm_test/pgm.json`
    - `tests/a.out`
    - `fortran_parsing_results.txt` (actually this seems to be from @pauldhein  - is it ok to remove?)
(**Note:** To keep a file on your local computer while removing it from the git repo, you can do 
`git rm --cached <filename>`)
2. Make a new directory under `delphi/tests` called `test_data` and move the joshua_toy_examples folder into it, as well as the other data files like `crop_yield.f`, `CROP_YIELD_DATA.INP`
3. Clean up any other stale files under `delphi/program_analysis/autoTranslate/tests`
  - Keep only the most up-to-date version of `crop_yield.f`, and remove other versions, like 
    - `tests/crop_yield_new.f`
    - `tests/crop_yield_return.f`
    - `tests/crop_yield_2.f`
    - `tests/latest_file.f`
    - `tests/file_crop.f`
  - The file `tests/orgn` seems like a binary file - is it needed? If not, remove it.
  - Are the files in `tests/scraps` needed? 
    - If yes, rename `scraps` to something more descriptive.
    - If not, remove the `scraps` folder.
4. Going forward, you may want to take advantage of Python's literal string interpolation feature (f-strings: https://www.python.org/dev/peps/pep-0498/) instead of str.format() for readability - you can see a few examples that I've done in pyFileTranslate.py
5. Also, in the future, you can replace some if-else cascades with a more readable dictionary lookup - see the refactoring in line 231 of pyFileTranslate.py
6. Rename `pyFileTranslate.py` to `pyTranslate.py` (do `git mv pyFileTranslate.py pyTranslate.py`)

**Comment by adarshp** [comment]
@pratikbhd Thanks for the cleanup and refactoring! I just pushed a few small cosmetic changes - basically I moved some of the comments describing functions into docstrings, which have native support (basically, they get stored in the __doc__ attribute of the function object) and can be readily recognized and compiled into API documentation using Sphinx (such as https://delphi.readthedocs.io/en/latest/delphi.program_analysis.html#module-delphi.program_analysis.printAst)

@skdebray Thanks for writing/pushing the f2py-pp.py script! The test coverage is measured using this tool: https://coverage.readthedocs.io/en/v4.5.x/.

The way to increase the coverage for f2py-pp.py is by writing unit/integration tests for it in the delphi/tests folder. You can take a look at the currently existing test code there to see some examples (I am by no means a testing virtuoso, so be prepared to see test code that's pretty rough around the edges...).

To run a particular test module, say `test_f2py-pp.py`, you can do 

```
pytest test_f2py-pp.py
```

To have the test session capture the output of print statements, run pytest with the -s flag:

```
pytest -s test_f2py-pp.py
```

**Comment by adarshp** [comment]
Also, @skdebray - I would like to merge this branch into master once @pratikbhd adds the comment at the top of `pyTranslate.py` - is that ok with you?

**Comment by skdebray** [comment]
@adarshp Once @pratikbhd has added in the top-level comment I'm fine with merging this in.

I have a question about adding coverage tests for the f2py-pp.py script: the delphi/tests directory you mentioned has several *.py files but the input to f2py-pp.py is a fortran file.  Do I just drop these fortran files into the delphi/tests directory?

**Comment by skdebray** [comment]
@adarshp never mind my previous question about testing f2py-pp.py.  You had addressed my question in your previous comment, I just didn't read carefully.  Thanks, and sorry. :)

**Comment by adarshp** [comment]
@skdebray No worries!

**Comment by pratikbhd** [comment]
@skdebray and @adarshp - Just added the top-level script information.

**Comment by adarshp** [comment]
@pratikbhd Excellent, thanks! I'm going to merge this branch and then delete it - in general, I'd like to have short-lived feature branches whenever possible...so make sure to do the following after I merge this branch:

```
git checkout master
git pull
git branch -D program_analysis
```

Then when you are implementing a new feature, just do

```
git checkout -b <branch-name>
```

where <branch-name> should reflect the feature you are implementing. Then when you think it's ready to merge, just submit a pull request for it.

**Comment by pratikbhd** [comment]
@adarshp Thanks! I'll follow this method moving forward.

**Pull request merged**